docs/src/main/asciidoc/security-openid-connect-client-registration.adoc

```diff
@@ -0,0 +1,332 @@
+
+You start with declaring one or more OIDC client registration configurations, for example:
+
+[source,properties]
+----
+# Default OIDC client registration which auto-discovers a standard client registration endpoint.
+# It does not require an initial registration token.
+
+quarkus.oidc-client-registration.auth-server-url=${quarkus.oidc.auth-server-url}
+quarkus.oidc-client-registration.metadata.client-name=Default Client
+quarkus.oidc-client-registration.metadata.redirect-uri=http://localhost:8081/protected
+
+# Named OIDC client registration which configures a registration endpoint path: 
+# It require an initial registration token for a client registration to succeed.
+
+quarkus.oidc-client-registration.tenant-client.registration-path=${quarkus.oidc.auth-server-url}/clients-registrations/openid-connect
+quarkus.oidc-client-registration.tenant-client.metadata.client-name=Tenant Client
+quarkus.oidc-client-registration.tenant-client.metadata.redirect-uri=http://localhost:8081/protected/tenant
+quarkus.oidc-client-registration.initial-token=${initial-registration-token}
+----
+
+The above configuration will lead to two new client registrations created in your OIDC provider.
+
+You or may not need to acquire an initial registration access token. If you don't, then you will have to enable one or more client registration policies in your OIDC provider's dashboard. For example, see https://www.keycloak.org/docs/latest/securing_apps/#_client_registration_policies[Keycloak client registration policies].   
+
+The next step is to inject either `quarkus.oidc.client.registration.OidcClientRegistration` if only a single default client registration is done, or `quarkus.oidc.client.registration.OidcClientRegistrations` if more than one registration is configured, and use metadata of the clients registered with these configurations.
+
+For example:
+
+[source,java]
+----
+package io.quarkus.it.keycloak;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Optional;
+
+import jakarta.enterprise.event.Observes;
+import jakarta.inject.Inject;
+import jakarta.inject.Singleton;
+
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+
+import io.quarkus.oidc.OidcRequestContext;
+import io.quarkus.oidc.OidcTenantConfig;
+import io.quarkus.oidc.OidcTenantConfig.ApplicationType;
+import io.quarkus.oidc.TenantConfigResolver;
+import io.quarkus.oidc.client.registration.ClientMetadata;
+import io.quarkus.oidc.client.registration.OidcClientRegistration;
+import io.quarkus.oidc.client.registration.OidcClientRegistrationConfig;
+import io.quarkus.oidc.client.registration.OidcClientRegistrations;
+import io.quarkus.oidc.client.registration.RegisteredClient;
+import io.smallrye.mutiny.Uni;
+import io.vertx.ext.web.RoutingContext;
+
+@Singleton
+public class CustomTenantConfigResolver implements TenantConfigResolver {
+
+    @Inject
+    OidcClientRegistration clientReg;
+
+    @Inject
+    OidcClientRegistrations clientRegs;
```

Comment:
What if we only have `OidcClientRegistration` and use a qualifier to inject a specific client registration? In case of injecting multiple instances, devs could use an injection point of  `Instance` ?

Reply:
@pedroigor Definitely we can do it, I've been thinking about doing it in phase 2. similarly to what was done for OidcClient.
Note `OidcClientRegistrations` covers 2 cases:
* Get access to individual `OidcClientRegistration` configurations done in `application.properties` and indeed, here, we can evolve it further to allow named injections, I did not do it straight away to keep the changes as minimum as possible and let the API settle a bit
* Accept new `OidcClientRegistration` configurations dynamically, for example, like in a case mentioned by @calvernaz, where a 3rd party client submits Client metadata to Quarkus which then initiates actual registrations to new realms etc